README.rst
.. README.rst
.. Copyright (c) 2013-2019 Pablo Acosta-Serafini
.. See LICENSE for details

.. image:: https://badge.fury.io/py/pexdoc.svg
    :target: https://pypi.org/project/pexdoc
    :alt: PyPI version

.. image:: https://img.shields.io/pypi/l/pexdoc.svg
    :target: https://pypi.org/project/pexdoc
    :alt: License

.. image:: https://img.shields.io/pypi/pyversions/pexdoc.svg
    :target: https://pypi.org/project/pexdoc
    :alt: Python versions supported

.. image:: https://img.shields.io/pypi/format/pexdoc.svg
    :target: https://pypi.org/project/pexdoc
    :alt: Format

|

.. image::
    https://dev.azure.com/pmasdev/pexdoc/_apis/build/status/pmacosta.pexdoc?branchName=master
    :target: https://dev.azure.com/pmasdev/pexdoc/_build?definitionId=3&_a=summary
    :alt: Continuous integration test status

.. image::
    https://img.shields.io/azure-devops/coverage/pmasdev/pexdoc/5.svg
    :target: https://dev.azure.com/pmasdev/pexdoc/_build?definitionId=5&_a=summary
    :alt: Continuous integration test coverage

.. image::
    https://readthedocs.org/projects/pip/badge/?version=stable
    :target: https://pip.readthedocs.io/en/stable/?badge=stable
    :alt: Documentation status

|

Description
===========

.. role:: bash(code)
	:language: bash

.. [[[cog
.. import os, sys, pmisc, docs.support.requirements_to_rst
.. file_name = sys.modules['docs.support.requirements_to_rst'].__file__
.. mdir = os.path.join(os.path.realpath(
..    os.path.dirname(os.path.dirname(os.path.dirname(file_name)))), 'pypkg'
.. )
.. docs.support.requirements_to_rst.def_links(cog)
.. ]]]
.. _Astroid: https://bitbucket.org/logilab/astroid
.. _Cog: https://nedbatchelder.com/code/cog
.. _Coverage: https://coverage.readthedocs.io
.. _Decorator: https://decorator.readthedocs.io
.. _Docutils: http://docutils.sourceforge.net/docs
.. _Funcsigs: https://pypi.org/project/funcsigs
.. _Mock: https://docs.python.org/3/library/unittest.mock.html
.. _Pmisc: http://pmisc.readthedocs.org
.. _Ptrie: http://ptrie.readthedocs.org
.. _PyContracts: https://andreacensi.github.io/contracts
.. _Pylint: https://www.pylint.org
.. _Py.test: http://pytest.org
.. _Pytest-coverage: https://pypi.org/project/pytest-cov
.. _Pytest-pmisc: http://pytest-pmisc.readthedocs.org
.. _Pytest-xdist: https://pypi.org/project/pytest-xdist
.. _Sphinx: http://sphinx-doc.org
.. _ReadTheDocs Sphinx theme: https://github.com/rtfd/sphinx_rtd_theme
.. _Inline Syntax Highlight Sphinx Extension:
   https://bitbucket.org/klorenz/sphinxcontrib-inlinesyntaxhighlight
.. _Shellcheck Linter Sphinx Extension:
   https://pypi.org/project/sphinxcontrib-shellcheck
.. _Tox: https://testrun.org/tox
.. _Virtualenv: https://docs.python-guide.org/dev/virtualenvs
.. [[[end]]]

This package provides a light framework that can be used to automatically
generate exceptions documentation marked up in
`reStructuredText <http://docutils.sourceforge.net/rst.html>`_.

The exdoc module details how to register exceptions, how to
traced them and how to generate their documentation

Interpreter
===========

The package has been developed and tested with Python 2.7, 3.5, 3.6 and 3.7
under Linux (Debian, Ubuntu), Apple macOS and Microsoft Windows

Installing
==========

.. code-block:: bash

	$ pip install pexdoc

Documentation
=============

Available at `Read the Docs <https://pexdoc.readthedocs.io>`_

Contributing
============

1. Abide by the adopted `code of conduct
   <https://www.contributor-covenant.org/version/1/4/code-of-conduct>`_

2. Fork the `repository <https://github.com/pmacosta/pexdoc>`_ from
   GitHub and then clone personal copy [#f1]_:

    .. code-block:: bash

        $ github_user=myname
        $ git clone --recursive \
              https://github.com/"${github_user}"/pexdoc.git
        Cloning into 'pexdoc'...
        ...
        $ cd pexdoc
        $ export PEXDOC_DIR=${PWD}

3. Install the project's Git hooks and build the documentation. The pre-commit
   hook does some minor consistency checks, namely trailing whitespace and
   `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ compliance via
   Pylint. Assuming the directory to which the repository was cloned is
   in the :bash:`$PEXDOC_DIR` shell environment variable:

	.. code-block:: bash

		$ "${PEXDOC_DIR}"/pypkg/complete-cloning.sh
                Installing Git hooks
                Building pexdoc package documentation
                ...

4. Ensure that the Python interpreter can find the package modules
   (update the :bash:`$PYTHONPATH` environment variable, or use
   `sys.paths() <https://docs.python.org/3/library/sys.html#sys.path>`_,
   etc.)

	.. code-block:: bash

		$ export PYTHONPATH=${PYTHONPATH}:${PEXDOC_DIR}

5. Install the dependencies (if needed, done automatically by pip):

    .. [[[cog
    .. import docs.support.requirements_to_rst
    .. docs.support.requirements_to_rst.proc_requirements(cog)
    .. ]]]


    * `Astroid`_ (1.6.0 or newer)

    * `Cog`_ (2.5.1 or newer)

    * `Coverage`_ (4.4.2 or newer)

    * `Decorator`_ (4.2.1 or newer)

    * `Docutils`_ (0.14 or newer)

    * `Funcsigs`_ (1.0.2 or newer)

    * `Inline Syntax Highlight Sphinx Extension`_ (0.2 or newer)

    * `Mock`_ (2.0.0 or newer)

    * `Pmisc`_ (1.5.5 or newer)

    * `Ptrie`_ (1.1.3 or newer)

    * `Py.test`_ (3.3.2 or newer)

    * `PyContracts`_ (1.8.2 or newer)

    * `Pylint`_ (1.8.1 or newer)

    * `Pytest-coverage`_ (2.5.1 or newer)

    * `Pytest-pmisc`_ (1.0.6 or newer)

    * `Pytest-xdist`_ (optional, 1.22.0 or newer)

    * `ReadTheDocs Sphinx theme`_ (0.2.4 or newer)

    * `Shellcheck Linter Sphinx Extension`_ (1.0.5 or newer)

    * `Sphinx`_ (1.6.6 or newer)

    * `Tox`_ (2.9.1 or newer)

    * `Virtualenv`_ (15.1.0 or newer)

    .. [[[end]]]

6. Implement a new feature or fix a bug

7. Write a unit test which shows that the contributed code works as expected.
   Run the package tests to ensure that the bug fix or new feature does not
   have adverse side effects. If possible achieve 100% code and branch
   coverage of the contribution. Thorough package validation
   can be done via Tox and Py.test:

	.. code-block:: bash

            $ tox
            GLOB sdist-make: .../pexdoc/setup.py
            py26-pkg inst-nodeps: .../pexdoc/.tox/dist/pexdoc-...zip

   `Setuptools <https://bitbucket.org/pypa/setuptools>`_ can also be used
   (Tox is configured as its virtual environment manager):

	.. code-block:: bash

	    $ python setup.py tests
            running tests
            running egg_info
            writing requirements to pexdoc.egg-info/requires.txt
            writing pexdoc.egg-info/PKG-INFO
            ...

   Tox (or Setuptools via Tox) runs with the following default environments:
   ``py27-pkg``, ``py35-pkg``, ``py36-pkg`` and ``py37-pkg`` [#f3]_. These use
   the 2.7, 3.5, 3.6 and 3.7 interpreters, respectively, to test all code in the
   documentation (both in Sphinx ``*.rst`` source files and in docstrings), run
   all unit tests, measure test coverage and re-build the exceptions
   documentation. To pass arguments to Py.test (the test runner) use a double
   dash (``--``) after all the Tox arguments, for example:

	.. code-block:: bash

	    $ tox -e py27-pkg -- -n 4
            GLOB sdist-make: .../pexdoc/setup.py
            py27-pkg inst-nodeps: .../pexdoc/.tox/dist/pexdoc-...zip
            ...

   Or use the :code:`-a` Setuptools optional argument followed by a quoted
   string with the arguments for Py.test. For example:

	.. code-block:: bash

	    $ python setup.py tests -a "-e py27-pkg -- -n 4"
            running tests
            ...

   There are other convenience environments defined for Tox [#f3]_:

    * ``py27-repl``, ``py35-repl``, ``py36-repl`` and ``py37-repl`` run the 2.7,
      3.5, 3.6 or 3.7 REPL, respectively, in the appropriate virtual
      environment. The ``pexdoc`` package is pip-installed by Tox when the
      environments are created.  Arguments to the interpreter can be passed in
      the command line after a double dash (``--``)

    * ``py27-test``, ``py35-test``, ``py36-test`` and ``py37-test`` run py.test
      using the Python 2.7, 3.5, Python 3.6 or Python 3.7 interpreter,
      respectively, in the appropriate virtual environment. Arguments to py.test
      can be passed in the command line after a double dash (``--``) , for
      example:

	.. code-block:: bash

	    $ tox -e py36-test -- -x test_pexdoc.py
            GLOB sdist-make: [...]/pexdoc/setup.py
            py36-test inst-nodeps: [...]/pexdoc/.tox/dist/pexdoc-1.1rc1.zip
            py36-test installed: -f file:[...]
            py36-test runtests: PYTHONHASHSEED='1264622266'
            py36-test runtests: commands[0] | [...]py.test -x test_pexdoc.py
            ===================== test session starts =====================
            platform linux -- Python 3.6.4, pytest-3.3.1, py-1.5.2, pluggy-0.6.0
            rootdir: [...]/pexdoc/.tox/py36/share/pexdoc/tests, inifile: pytest.ini
            plugins: xdist-1.21.0, forked-0.2, cov-2.5.1
            collected 414 items
            ...

    * ``py27-cov``, ``py35-cov``, ``py36-cov`` and ``py37-cov`` test code and
      branch coverage using the 2.7, 3.5, 3.6 or 3.7 interpreter, respectively,
      in the appropriate virtual environment. Arguments to py.test can be passed
      in the command line after a double dash (``--``). The report can be found
      in
      :bash:`${PEXDOC_DIR}/.tox/py[PV]/usr/share/pexdoc/tests/htmlcov/index.html`
      where ``[PV]`` stands for ``27``, ``35``, ``36`` or ``37`` depending on
      the interpreter used

8. Verify that continuous integration tests pass. The package has continuous
   integration configured for Linux, Apple macOS and Microsoft Windows (all via
   `Azure DevOps <https://dev.azure.com/pmasdev>`_) Aggregation/cloud code
   coverage is configured via `Codecov <https://codecov.io>`_. It is assumed
   that the Codecov repository upload token in the build is stored in the
   :bash:`$(codecovToken)` environment variable (securely defined in the
   pipeline settings page).

9. Document the new feature or bug fix (if needed). The script
   :bash:`${PEXDOC_DIR}/pypkg/build_docs.py` re-builds the whole package
   documentation (re-generates images, cogs source files, etc.):

	.. [[[cog pmisc.ste('build_docs.py -h', 0, mdir, cog.out) ]]]

	.. code-block:: bash

	    $ ${PKG_BIN_DIR}/build_docs.py -h
	    usage: build_docs.py [-h] [-d DIRECTORY] [-r]
	                         [-n NUM_CPUS] [-t]

	    Build pexdoc package documentation

	    optional arguments:
	      -h, --help            show this help message and exit
	      -d DIRECTORY, --directory DIRECTORY
	                            specify source file directory
	                            (default ../pexdoc)
	      -r, --rebuild         rebuild exceptions documentation.
	                            If no module name is given all
	                            modules with auto-generated
	                            exceptions documentation are
	                            rebuilt
	      -n NUM_CPUS, --num-cpus NUM_CPUS
	                            number of CPUs to use (default: 1)
	      -t, --test            diff original and rebuilt file(s)
	                            (exit code 0 indicates file(s) are
	                            identical, exit code 1 indicates
	                            file(s) are different)

	.. [[[end]]]

.. rubric:: Footnotes

.. [#f1] All examples are for the `bash <https://www.gnu.org/software/bash/>`_
   shell

.. [#f2] It is assumed that all the Python interpreters are in the executables
   path. Source code for the interpreters can be downloaded from Python's main
   `site <https://www.python.org/downloads/>`_

.. [#f3] Tox configuration largely inspired by
   `Ionel's codelog <https://blog.ionelmc.ro/2015/04/14/
   tox-tricks-and-patterns/>`_


License
=======

The MIT License (MIT)

Copyright (c) 2013-2019 Pablo Acosta-Serafini

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.rst
.. CHANGELOG.rst
.. Copyright (c) 2013-2019 Pablo Acosta-Serafini
.. See LICENSE for details

Changelog
=========

* 1.1.0 [2018-01-19]: Dropped support for Python interpreter versions 2.6, 3.3
  and 3.4. Updated dependencies versions to their current versions

* 1.0.9 [2017-02-10]: Package build enhancements and fixes

* 1.0.8 [2017-02-07]: Python 3.6 support

* 1.0.7 [2017-01-02]: Suppressed deprecation warnings for
  Python 2.6

* 1.0.6 [2016-06-13]: Fixed Windows continuous integration
  failure

* 1.0.5 [2016-06-11]: Minor documentation build bug fix

* 1.0.4 [2016-05-13]: Documentation update

* 1.0.3 [2016-05-11]: Documentation update

* 1.0.2 [2016-05-11]: Documentation update

* 1.0.1 [2016-05-11]: Documentation update

* 1.0.0 [2016-05-01]: Final release of 1.0.0 branch

* 1.0.0rc1 [2016-05-01]: Initial commit, forked a subset from putil PyPI
  package


sbin/build_docs.py
#!/usr/bin/env python
# build_docs.py
# Copyright (c) 2013-2017 Pablo Acosta-Serafini
# See LICENSE for details
# pylint: disable=C0111,C0411,C0413,E0611,F0401,R0912,R0914,R0915,W0141

# Standard library imports
from __future__ import print_function
import argparse
import datetime
import difflib
import glob
import multiprocessing
import os
import re
import shutil
import sys
# PyPI imports
from cogapp import Cog
# Intra-package imports
import sbin.functions
try:
    from sbin.refresh_moddb import refresh_moddb
except ImportError:
    def refresh_moddb():
        pass
try:
    from sbin.build_moddb import build_moddb
except ImportError:
    def build_moddb():
        pass


###
# Global variables
###
PKG_NAME = 'pexdoc'
VALID_MODULES = [PKG_NAME]
PKG_SUBMODULES = []


###
# Functions
###
def build_pkg_docs(args):
    """ Build documentation """
    debug = False
    retcode = 0
    pkg_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    src_dir = args.directory
    cog_exe = which('cog.py')
    if not cog_exe:
        raise RuntimeError('cog binary could not be found')
    os.environ['NOPTION'] = (
        '-n {0}'.format(args.num_cpus) if args.num_cpus > 1 else ''
    )
    rebuild = args.rebuild
    test = args.test
    tracer_dir = os.path.join(pkg_dir, 'docs', 'support')
    os.environ['TRACER_DIR'] = tracer_dir
    # Processing
    del_pkl_files(test, tracer_dir)
    print('Rebuilding documentation')
    if debug:
        print('Python: {0}'.format(sys.executable))
        print('PATH: {0}'.format(os.environ['PATH']))
        print('PYTHONPATH: {0}'.format(os.environ['PYTHONPATH']))
        print('Cog: {0}'.format(cog_exe))
        print('sbin.functions: {0}'.format(sbin.functions.__file__))
        print(
            'sbin.functions.subprocess: {0}'.format(
                sbin.functions.subprocess.__file__
            )
        )
    if rebuild or test:
        refresh_moddb()
        print_cyan(
            'Rebuilding exceptions documentation{0}'.format(
                ' (test mode)' if test else ''
            ),
        )
        start_time = datetime.datetime.today()
        tmp_retcode = rebuild_module_doc(
            test, src_dir, tracer_dir, cog_exe, debug
        )
        retcode = tmp_retcode if not retcode else retcode
        stop_time = datetime.datetime.today()
        print(
            'Elapsed time: {0}'.format(
                elapsed_time_string(start_time, stop_time)
            )
        )
        build_moddb()
    generate_top_level_readme(pkg_dir)
    print('Inserting files into docstrings')
    insert_files_in_rsts(pkg_dir, cog_exe)
    print('Generating HTML output')
    shutil.rmtree(os.path.join(pkg_dir, 'docs', '_build'), ignore_errors=True)
    cwd = os.getcwd()
    os.chdir(os.path.join(pkg_dir, 'docs'))
    sbin.functions.shcmd(
        [
            'sphinx-build',
            '-b', 'html',
            '-d', os.path.join('_build', 'doctrees'),
            '-W', '.',
            os.path.join('_build', 'html')
        ],
        'Error building Sphinx documentation',
        async_stdout=True
    )
    # Copy built documentation to its own directory
    # dest_dir = os.path.join(pkg_dir, 'docs', 'html')
    # src_dir = os.path.join(pkg_dir, 'docs', '_build', 'html')
    # shutil.rmtree(dest_dir, ignore_errors=True)
    # shutil.copytree(src_dir, dest_dir)
    os.chdir(cwd)
    return retcode


def copy_file(src, dest):
    """ Copy file (potentially overwriting existing file) """
    try:
        os.remove(dest)
    except OSError:
        pass
    shutil.copy(src, dest)


def del_file(fname):
    """ Delete file """
    try:
        os.remove(fname)
    except OSError:
        pass


def del_pkl_files(test, tracer_dir):
    """ Delete all old pickle files """
    if test:
        pkl_files = (glob.glob(os.path.join(tracer_dir, '*.pkl')))
        for pkl_file in pkl_files:
            os.remove(pkl_file)


def diff(file1, file2):
    """ Diff two files """
    with open(file1, 'r') as fobj1:
        flines1 = [item.rstrip() for item in fobj1.readlines()]
    with open(file2, 'r') as fobj2:
        flines2 = [item.rstrip() for item in fobj2.readlines()]
    return list(
        difflib.unified_diff(flines1, flines2, fromfile=file1, tofile=file2)
    )


def elapsed_time_string(start_time, stop_time):
    """
    Returns a formatted string with the elapsed time between two time points
    """
    delta_time = stop_time-start_time
    tot_seconds = int(
        (
            delta_time.microseconds+
            (delta_time.seconds+delta_time.days*24*3600)*10**6
        )
        /
        10**6
    )
    years, remainder = divmod(tot_seconds, 365*24*60*60)
    months, remainder = divmod(remainder, 30*24*60*60)
    days, remainder = divmod(remainder, 24*60*60)
    hours, remainder = divmod(remainder, 60*60)
    minutes, seconds = divmod(remainder, 60)
    token_iter = zip(
        [years, months, days, hours, minutes, seconds],
        ['year', 'month', 'day', 'hour', 'minute', 'second']
    )
    ret_list = [
        '{token} {token_name}{plural}'.format(
            token=num,
            token_name=desc,
            plural='s' if num > 1 else ''
        ) for num, desc in token_iter if num > 0
    ]
    if len(ret_list) == 0:
        return 'None'
    elif len(ret_list) == 1:
        return ret_list[0]
    elif len(ret_list) == 2:
        return ret_list[0]+' and '+ret_list[1]
    else:
        return (', '.join(ret_list[0:-1]))+' and '+ret_list[-1]


def insert_files_in_rsts(pkg_dir, cog_exe):
    """ Cog-insert source files in Sphinx files """
    fnames = [
        os.path.join(pkg_dir, 'docs', 'README.rst'),
        os.path.join(pkg_dir, 'README.rst'),
    ]
    print('Inserting source files in documentation files')
    for fname in fnames:
        print('   Processing file {0}'.format(fname))
        retcode = Cog().main(
            [
                cog_exe,
                '-e', '-x', '-o', fname+'.tmp', fname
            ]
        )
        if retcode:
            raise RuntimeError(
		        'Error deleting insertion of source files in '
		        'documentation file {0}'.format(fname)
            )
        retcode = Cog().main(
            [
                cog_exe,
                '-e', '-o', fname+'.tmp', fname
            ]
        )
        if retcode:
            raise RuntimeError(
		        'Error inserting source files in '
		        'docstrings in module {0}'.format(fname)
            )
        move_file(fname+'.tmp', fname)


def move_file(src, dest):
    """ Copy file (potentially overwriting existing file) """
    try:
        os.remove(dest)
    except OSError:
        pass
    shutil.move(src, dest)


def pcolor(text, color, indent=0):
    r"""
    Returns a string that once printed is colorized (copied from pmisc module)

    :param text: Text to colorize
    :type  text: string

    :param  color: Color to use, one of :code:`'black'`, :code:`'red'`,
                   :code:`'green'`, :code:`'yellow'`, :code:`'blue'`,
                   :code:`'magenta'`, :code:`'cyan'`, :code:`'white'` or
                   :code:`'none'` (case insensitive)
    :type   color: string

    :param indent: Number of spaces to prefix the output with
    :type  indent: integer

    :rtype: string

    :raises:
     * RuntimeError (Argument \`color\` is not valid)

     * RuntimeError (Argument \`indent\` is not valid)

     * RuntimeError (Argument \`text\` is not valid)

     * ValueError (Unknown color *[color]*)
    """
    esc_dict = {
        'black':30, 'red':31, 'green':32, 'yellow':33, 'blue':34, 'magenta':35,
        'cyan':36, 'white':37, 'none':-1
    }
    if not isinstance(text, str):
        raise RuntimeError('Argument `text` is not valid')
    if not isinstance(color, str):
        raise RuntimeError('Argument `color` is not valid')
    if not isinstance(indent, int):
        raise RuntimeError('Argument `indent` is not valid')
    color = color.lower()
    if color not in esc_dict:
        raise ValueError('Unknown color {color}'.format(color=color))
    if esc_dict[color] != -1:
        return (
            '\033[{color_code}m{indent}{text}\033[0m'.format(
                color_code=esc_dict[color],
                indent=' '*indent,
                text=text
            )
        )
    return '{indent}{text}'.format(indent=' '*indent, text=text)


def print_diff(tlist, indent=3):
    """ Pretty prints file differences """
    ret = []
    ret.append((indent*' ')+tlist[0][1:-2])
    ret.append((indent*' ')+tlist[1][1:-2])
    for line in tlist[2:]:
        ret.append((indent*' ')+str(line.rstrip()))
    return '\n'.join(ret)


def print_cyan(text):
    """ Print text to STDOUT in cyan color """
    print(pcolor(text, 'cyan'))


def print_green(text):
    """ Print text to STDOUT in green color """
    print(pcolor(text, 'green'))


def print_red(text):
    """ Print text to STDOUT in red color """
    print(pcolor(text, 'red'))


def rebuild_module_doc(test, src_dir, tracer_dir, cog_exe, debug):
    # pylint: disable=R0913
    retcode = 0
    pkl_dir = tracer_dir
    submodules = PKG_SUBMODULES
    for submodule in submodules:
        smf = os.path.join(src_dir, submodule+'.py')
        pkl_file = os.path.join(pkl_dir, submodule+'.pkl')
        print_cyan('Processing module {0}'.format(submodule))
        orig_file = smf+'.orig'
        if test:
            shutil.copy(smf, orig_file)
        stdout = sbin.functions.shcmd(
            [sys.executable, cog_exe, '-e', '-o', smf+'.tmp', smf],
            (
                'Error generating exceptions documentation '
                'in module {0}'.format(smf)
            ),
            async_stdout=debug
        )
        move_file(smf+'.tmp', smf)
        if test:
            diff_list = diff(smf, orig_file)
            if not diff_list:
                print_green('   File {0} identical from original'.format(smf))
                del_file(pkl_file)
            else:
                print_red('   File {0} differs from original'.format(smf))
                print('   Differences:')
                print(print_diff(diff_list))
                copy_file(smf, smf+'.error')
                retcode = 1
            move_file(orig_file, smf)
        else:
            print(stdout)
            del_file(pkl_file)
    return retcode


def generate_top_level_readme(pkg_dir):
    """
    Remove Sphinx-specific cross-references from top-level README.rst file,
    they are not rendered by either Bitbucket or GitHub
    """
    # pylint: disable=W0212
    docs_dir = os.path.abspath(os.path.join(pkg_dir, 'docs'))
    fname = os.path.join(docs_dir, 'README.rst')
    print('Generating top-level README.rst file')
    with open(fname, 'r') as fobj:
        lines = [item.rstrip() for item in fobj.readlines()]
    ref1_regexp = re.compile('.*:py:mod:`(.+) <'+PKG_NAME+'.(.+)>`.*')
    ref2_regexp = re.compile('.*:py:mod:`'+PKG_NAME+'.(.+)`.*')
    ref3_regexp = re.compile(r'.*:ref:`(.+?)(\s+<.+>)*`.*')
    rst_cmd_regexp = re.compile('^\\s*.. \\S+::.*')
    indent_regexp = re.compile('^(\\s*)\\S+')
    ret = []
    autofunction = False
    literalinclude = False
    for line in lines:
        match1 = ref1_regexp.match(line)
        match2 = ref2_regexp.match(line)
        match3 = ref3_regexp.match(line)
        if autofunction:
            match = indent_regexp.match(line)
            if (not match) or (match and len(match.group(1)) == 0):
                autofunction = False
                ret.append(line)
        elif literalinclude:
            if line.lstrip().startswith(':lines:'):
                literalinclude = False
                lrange = line.lstrip().replace(':lines:', '').strip()
                tstr = (
                    '.. docs.support.incfile.incfile(\n'
                    '..     "{0}",\n'
                    '..     cog.out,\n'
                    '..     "{1}",\n'
                    '..     None\n'
                    '.. )'
                )
                ret.append('.. [[[cog')
                ret.append('.. import docs.support.incfile')
                ret.append(tstr.format(os.path.basename(fname), lrange))
                ret.append('.. ]]]')
                ret.append('.. [[[end]]]')
        elif match1:
            # Remove cross-references
            label = match1.group(1)
            mname = match1.group(2)
            line = line.replace(
                ':py:mod:`{label} <'.format(label=label)+
                PKG_NAME+'.{mname}>`'.format(mname=mname),
                label
            )
            ret.append(line)
        elif match2:
            # Remove cross-references
            mname = match2.group(1)
            line = line.replace(
                ':py:mod:`'+PKG_NAME+'.{mname}`'.format(mname=mname), mname
            )
            ret.append(line)
        elif match3:
            # Remove cross-references
            mname = match3.group(1)
            target = match3.group(2)
            line = line.replace(
                ':ref:`{mname}{target}`'.format(
                    mname=mname,
                    target='' if target is None else target
                ), mname
            )
            ret.append(line)
        elif line.lstrip().startswith('.. literalinclude::'):
            fname = line.lstrip().replace('.. literalinclude::', '').strip()
            literalinclude = True
        elif line.lstrip().startswith('.. include::'):
            # Include files
            base_fname = line.split()[-1].strip()
            fname = os.path.basename(base_fname)
            # Do not include the change log, PyPI adds it at the end
            # of the README.rst file by default and in a hosted Git
            # repository there is a much more detailed built-in change
            # log in the commit message history
            if fname != 'CHANGELOG.rst':
                fname = os.path.join(docs_dir, base_fname)
                for inc_line in sbin.functions._readlines(fname):
                    comment = inc_line.lstrip().startswith('.. ')
                    if ((not comment)
                       or (comment and rst_cmd_regexp.match(inc_line))):
                        ret.append(inc_line.rstrip())
        elif line.lstrip().startswith('.. autofunction::'):
            # Remove auto-functions, PyPI reStructuredText parser
            # does not appear to like it
            autofunction = True
        else:
            ret.append(line)
    fname = os.path.join(pkg_dir, 'README.rst')
    with open(fname, 'w') as fobj:
        fobj.write('\n'.join(ret))
    # Check that generated file produces HTML version without errors
    sbin.functions.shcmd(
        ['rst2html.py', '--exit-status=3', '-v', fname],
        'Error validating top-level README.rst HTML conversion',
    )


def valid_dir(value):
    """ Argparse checked for directory argument """
    if not os.path.isdir(value):
        raise argparse.ArgumentTypeError(
            'directory {0} does not exist'.format(value)
        )
    return os.path.abspath(value)


def valid_num_cpus(value):
    """ Argparse checker for num_cpus argument """
    # pylint: disable=E1101
    try:
        value = int(value)
    except:
        raise argparse.ArgumentTypeError(
            'invalid positive integer value: {0}'.format(value)
        )
    if value < 1:
        raise argparse.ArgumentTypeError(
            'invalid positive integer value: {0}'.format(value)
        )
    max_cpus = multiprocessing.cpu_count()
    if value > max_cpus:
        raise argparse.ArgumentTypeError(
            'requested CPUs ({0}) greater than '
            'available CPUs ({1})'.format(value, max_cpus)
        )
    return value


def which(name):
    """ Search PATH for executable files with the given name """
    # Inspired by https://twistedmatrix.com/trac/browser/tags/releases/
    # twisted-8.2.0/twisted/python/procutils.py
    result = []
    path = os.environ.get('PATH', None)
    if path is None:
        return []
    for pdir in os.environ.get('PATH', '').split(os.pathsep):
        fname = os.path.join(pdir, name)
        if os.path.isfile(fname) and os.access(fname, os.X_OK):
            result.append(fname)
    return result[0] if result else None


if __name__ == "__main__":
    # pylint: disable=E0602
    PKG_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    PARSER = argparse.ArgumentParser(
        description='Build '+PKG_NAME+' package documentation'
    )
    PARSER.add_argument(
        '-d', '--directory',
        help='specify source file directory (default ../'+PKG_NAME+')',
        type=valid_dir,
        nargs=1,
        default=[os.path.join(PKG_DIR, PKG_NAME)]
    )
    PARSER.add_argument(
        '-r', '--rebuild',
        help=(
            'rebuild exceptions documentation. If no module name '
            'is given all modules with auto-generated exceptions '
            'documentation are rebuilt'
        ),
        action="store_true"
    )
    PARSER.add_argument(
        '-n', '--num-cpus',
        help='number of CPUs to use (default: 1)',
        type=valid_num_cpus,
        default=1
    )
    PARSER.add_argument(
        '-t', '--test',
        help=(
            'diff original and rebuilt file(s) (exit code 0 '
            'indicates file(s) are identical, exit code 1 '
            'indicates file(s) are different)'

        ),
        action="store_true"
    )
    ARGS = PARSER.parse_args()
    ARGS.directory = ARGS.directory[0]
    if ARGS.rebuild and (not ARGS.test):
        if sys.hexversion < 0x03000000: # pragma: no cover
            VAR = raw_input('Are you sure [Y/N]? ')
        else:
            VAR = input('Are you sure [Y/N]? ')
        if VAR.lower() != 'y':
            sys.exit(0)
    sys.exit(build_pkg_docs(ARGS))
